FirebaseRemoteConfig/Sources/RCNFetch.m

```diff
@@ -489,23 +494,33 @@ - (void)fetchWithUserProperties:(NSDictionary *)userProperties
     });
   };
 
-  if (gGlobalTestBlock) {
-    gGlobalTestBlock(fetcherCompletion);
+  // If there is a fake config, use that and skip the big block above that does the remote fetch.
+  if ([RCNFakeFetch active]) {
+    self->_settings.isFetchInProgress = NO;
+
+    NSDictionary<NSString *, id> *fakeConfig = [RCNFakeFetch get];
+    // Update config content to cache and DB.
+    if ([fakeConfig[@"state"] isEqualToString:RCNFetchResponseKeyStateUpdate]) {
+      NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
+      [formatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
+      self->_settings.lastETag = [formatter stringFromDate:[NSDate date]];
+    }
+    [_content updateConfigContentWithResponse:fakeConfig forNamespace:self->_FIRNamespace];
+    // Update experiments.
+    [_experiment
+        updateExperimentsWithResponse:fakeConfig[RCNFetchResponseKeyExperimentDescriptions]];
+    [self->_settings updateMetadataWithFetchSuccessStatus:YES];
+    completionHandler(FIRRemoteConfigFetchStatusSuccess, nil);
     return;
   }
+
   FIRLogDebug(kFIRLoggerRemoteConfig, @"I-RCN000061", @"Making remote config fetch.");
 
   NSURLSessionDataTask *dataTask = [self URLSessionDataTaskWithContent:compressedContent
                                                      completionHandler:fetcherCompletion];
   [dataTask resume];
 }
 
-+ (void)setGlobalTestBlock:(RCNConfigFetcherTestBlock)block {
-  FIRLogDebug(kFIRLoggerRemoteConfig, @"I-RCN000027",
-              @"Set global test block for NSSessionFetcher, it will not fetch from server.");
-  gGlobalTestBlock = [block copy];
```

Comment:
I feel like using a test block is simpler and doesn't take a lot of test code in production. And you can customize the block in test rather than having all the test logic in production code. You can also provide flexible return value (not just success) to test different scenarios.

This is used in a lot of the unit tests of remote config (search setGlobalTestBlock in tests), but I'm not sure the current state of how many of those tests are running but if we remove this, we might also have to adjust all those unit test as well. 

Another thing is do we want to fake fetch in integration test since in github network is available. isn't it better to actually enable fetch from prod server?

Reply:
The test block is only used in disabled unit tests that still need to be ported from RC v1 to v2.  This implementation allows the tests to be implemented without any knowledge of the library implementation details - just a simple API of managing a dictionary.

The swift_api tests are similar to these but do fetches from a prod server. This set of tests are to enable API tests that require midstream console changes.